hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/OMVolumeDeleteRequest.java

```diff
@@ -0,0 +1,200 @@
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .DeleteVolumeResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMResponse;
+import org.apache.hadoop.utils.db.cache.CacheKey;
+import org.apache.hadoop.utils.db.cache.CacheValue;
+
+
+/**
+ * Handles volume delete request.
+ */
+public class OMVolumeDeleteRequest extends OMClientRequest
+    implements OMVolumeRequest {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMVolumeDeleteRequest.class);
+
+  public OMVolumeDeleteRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long transactionLogIndex) {
+
+    DeleteVolumeRequest deleteVolumeRequest =
+        getOmRequest().getDeleteVolumeRequest();
+    Preconditions.checkNotNull(deleteVolumeRequest);
+
+    String volume = deleteVolumeRequest.getVolumeName();
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumVolumeDeletes();
+
+    OMResponse.Builder omResponse = OMResponse.newBuilder().setCmdType(
+        OzoneManagerProtocolProtos.Type.DeleteVolume).setStatus(
+        OzoneManagerProtocolProtos.Status.OK).setSuccess(true);
+
+    AuditLogger auditLogger = ozoneManager.getAuditLogger();
+    OzoneManagerProtocolProtos.UserInfo userInfo = getOmRequest().getUserInfo();
+
+    try {
+      // check Acl
+      if (ozoneManager.getAclsEnabled()) {
+        checkAcls(ozoneManager, OzoneObj.ResourceType.VOLUME,
+            OzoneObj.StoreType.OZONE, IAccessAuthorizer.ACLType.DELETE, volume,
+            null, null);
+      }
+    } catch (IOException ex) {
+      LOG.error("Volume deletion failed for volume:{}", volume, ex);
+      omMetrics.incNumVolumeDeleteFails();
+      auditLog(auditLogger, buildAuditMessage(OMAction.DELETE_VOLUME,
+          buildVolumeAuditMap(volume), ex, userInfo));
+      return new OMVolumeCreateResponse(null, null,
+          createErrorOMResponse(omResponse, ex));
+    }
+
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+
+    OmVolumeArgs omVolumeArgs = null;
+    String owner = null;
+
+    omMetadataManager.getLock().acquireVolumeLock(volume);
+    try {
+      owner = getVolumeInfo(omMetadataManager, volume).getOwnerName();
+    } catch (IOException ex) {
+      LOG.error("Volume deletion failed for volume:{}", volume, ex);
+      omMetrics.incNumVolumeDeleteFails();
+      auditLog(auditLogger, buildAuditMessage(OMAction.DELETE_VOLUME,
+          buildVolumeAuditMap(volume), ex, userInfo));
+      return new OMVolumeDeleteResponse(null, null, null,
+          createErrorOMResponse(omResponse, ex));
+    } finally {
+      omMetadataManager.getLock().releaseVolumeLock(volume);
+    }
+
+    // We cannot acquire user lock holding volume lock, so released volume
```

Comment:
This looks risky. Can the state of the world change between releasing the lock and reacquiring it?

Also I am thinking why do we need the user lock in the first place?

Reply:
From my understanding user-lock is for protecting userTable, where we maintain list of volumes for that particular user.

Reply:
What about the release/reacquire issue? Same issue exists in OMVolumeSetOwnerRequest.

Reply:
Example:
SetOwner Volume - V1 , old Owner ozone, newOwner hadoop
CreateVolume Volume -v7 owner-ozone

If we don't acquire the lock for oldOwner before reading from the table. We might see an issue. 

This is to protect above kind of scenario

Reply:
And holding lock for the user is to protect volumeList which we store for the user.

Reply:
Bharat - let's discuss. What I am mean is that release/reacquire pattern is usually a bad idea. It may be okay here, however it would be better to eliminate it.

Reply:
Okay..